tools/ci_build/exclude_unused_ops_and_types.py

```diff
@@ -0,0 +1,176 @@
+    '''rewrite provider registration file to exclude unused ops'''
+
+    for kernel_registration_file in provider_registration_paths:
+        if not os.path.isfile(kernel_registration_file):
+            raise ValueError('Kernel registration file {} does not exist'.format(kernel_registration_file))
+
+        log.info("Processing {}".format(kernel_registration_file))
+
+        backup_path = kernel_registration_file + '~'
+        shutil.move(kernel_registration_file, backup_path)
+
+        # read from backup and overwrite original with commented out lines for any kernels that are not required
+        with open(kernel_registration_file, 'w') as file_to_write:
+            processor = ExcludeOpsAndTypesRegistrationProcessor(required_operators,
+                                                                op_type_usage_manager,
+                                                                file_to_write)
+
+            op_registration_utils.process_kernel_registration_file(backup_path, processor)
+
+            if not processor.ok():
+                # error should have already been logged so just exit
+                sys.exit(-1)
+
+
+def _generate_cpp_defines(ort_root: str, op_type_usage_manager: OperatorTypeUsageManager):
+
+    # get the C++ code to insert
+    cpp_lines = op_type_usage_manager.get_cpp_entries() if op_type_usage_manager else None
+    if not cpp_lines:
+        return
+
+    target = os.path.join(ort_root, 'onnxruntime', 'core', 'providers', 'op_kernel_type_control_overrides.inc')
+    if not os.path.exists(target) or not os.path.isfile(target):
+        log.warning('Could not find {}. Skipping generation of C++ code to reduce the types supported by operators.'
+                    .format(target))
+        return
+
+    # copy existing content to use as input
+    src = target + '.tmp'
+    shutil.copyfile(target, src)
+
+    # find the insertion block and replace any existing content in it
+    inserted = False
+    with open(src, 'r') as input, open(target, 'w') as output:
+        inside_insertion_block = False
+        for line in input.readlines():
+            if '@@insertion_point_begin(allowed_types)@@' in line:
+                inside_insertion_block = True
+                output.write(line)
+                [output.write('{}\n'.format(code_line)) for code_line in cpp_lines]
+                inserted = True
+                continue
+            elif inside_insertion_block:
+                if '@@insertion_point_end(allowed_types)@@' in line:
+                    inside_insertion_block = False
+                else:
+                    # we ignore any old lines within the insertion block
+                    continue
+
+            output.write(line)
+
+    os.remove(src)
+
+    if not inserted:
+        raise RuntimeError('Insertion point was not found in {}'.format(target))
+
+    # future: how/where will we write global type limitations?
```

Comment:
>write global type limitations [](start = 32, length = 29)

the C++ code is currently assuming that all allowed type specifications will be in the same place

Reply:
still questions regarding where the global spec comes from (e.g. manually editing the .inc file, adding to the reduced build config, as a build.py param). it will be added to the .inc file though so that part of 'where' is resolved. will update the comment